Support switching anti-aliasing and multi-sampling from CVar

hosea1008 opened the pull request:
## What does this PR do?

There are 3 types of anti-aliasing method in the engine, MSAA, SMAA and TAA. But there are not a proper way to switch between them. So we introduce this PR to support switching between different anti-aliasing and multi-sampling. 

We think the AA and multi-sampling setting is a project-wise setting, so we put the switch in the project manager, one can set the AA and multi-sampling setting in the `Edit Project Settings` menu.

<img width="588" alt="捕获" src="https://github.com/o3de/o3de/assets/5073925/83fc7536-4feb-43cb-93cc-93622229b318">


The AA and multi-sampling setting can be hot loaded in the main editor when you change the settings in project manager. We also display the current AA and multi-sampling setting in the editor's display info.

<img width="244" alt="捕获2" src="https://github.com/o3de/o3de/assets/5073925/cc068c4b-9002-4625-a7db-d0c5a8d9b445">


## How was this PR tested?

Self tested, and this feature has been used in quite a lot of our real projects.

VickyAtAZ reviewed (commented):
I think the design of setting the AA and multisample count in project may not be a good decision. Here are some thoughts I have about the design:
1. The AA methods are mainly decided by the feature of the render pipeline which is part of Atom Common Feature gem and RPI doesn't have knowledge about that. 
2. Add Atom libraries as dependencies of ProjectManager just for a setting is overkill. 
3. Only allow change the AA or MS settings as project setting seems adding too much restrictions. It's common games may have options to switch these settings in their graphics settings in game. (Hot-reloading is only intended for editing environment. For a release launcher, we won't connect to AP so it won't be triggered )
4. I think we already can switch AA method and multi-sample state at runtime per render pipeline. (I don't remember the detail but I remember there were changes made to support that).

VickyAtAZ commented:
My thoughts on possible workflow:
Option 1. The AA setting and MS setting can be added as a level component.  The implementation of changing them would be done in the component's implementation `Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZStd::vector<AZ::Crc32>({ AZ_CRC_CE("Level") }))`
Option 2. Add cvars for AA and MS settings. Add cvar change handling for runtime value change. The Cvars and their handling functions can be added to BootstrapSystemComponent. (CVars are configurable via setreg file too). 
I prefer option 2 since it's involve minimum change and flexible.

DIFG-github commented:
I agree the most with option 2, since cvars are easy to find, edit, and it gives the easiest implementation to be able to switch while in editor without having a component in the level entity. 

I do agree with hosea1008 that as of right now this feature is needed because switching these settings is not user friendly at this time.

moudgils commented:
Agree with most oft he points made by Vicky. You are essentially coupling Project Manager to a rendering feature which is not good as so many things can go wrong. What if the rendering pipeline doesnt support the selected AA option? For example the MultiView pipeline does not support SMAA or TAA. I also prefer option 2. If we add AA support to cvar it will allow game developers to easily add a menu to allow end users (i.e gamers) to switch AA easily.

hosea1008 commented:
OK, Thanks for the suggestions, I will consider updating it in the following days

hosea1008 commented:
@VickyAtAZ @moudgils @DIFG-github Hi, now the anti-aliasing and multi-sample count is set by 2 CVars `r_antiAliasing` and `r_multiSample`, they can be updated during runtime. And I removed the dependency of project manager to Atom.

hosea1008 commented:
![捕获3](https://github.com/o3de/o3de/assets/5073925/0f9b9958-8809-4c1f-9c13-5fcaee58a7cc)
![捕获3](https://github.com/o3de/o3de/assets/5073925/f9095ea2-1ced-4069-9071-b982dfbd9f7e)
![捕获3](https://github.com/o3de/o3de/assets/5073925/68b41d5f-750b-4cce-bfb7-ae7ba4bbda22)

hosea1008 commented:
![捕获3](https://github.com/o3de/o3de/assets/5073925/0f9b9958-8809-4c1f-9c13-5fcaee58a7cc)
![捕获3](https://github.com/o3de/o3de/assets/5073925/f9095ea2-1ced-4069-9071-b982dfbd9f7e)
![捕获3](https://github.com/o3de/o3de/assets/5073925/68b41d5f-750b-4cce-bfb7-ae7ba4bbda22)

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -74,6 +74,45 @@ void cvar_r_renderPipelinePath_Changed(const AZ::CVarFixedString& newPipelinePat
     }
 }
 
+void cvar_r_antiAliasing_Changed(const AZ::CVarFixedString& newAtiAliasing)
```
Typo: newAntiAliasing

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.h:
```diff
@@ -70,6 +70,8 @@ namespace AZ
                 bool EnsureDefaultRenderPipelineInstalledForScene(AZ::RPI::ScenePtr scene, AZ::RPI::ViewportContextPtr viewportContext) override;
                 void SwitchRenderPipeline(const AZ::RPI::RenderPipelineDescriptor& newRenderPipelineDesc, AZ::RPI::ViewportContextPtr viewportContext) override;
+                void SwitchAntiAliasing(const AZStd::string& newAntiAliasing, AZ::RPI::ViewportContextPtr viewportContext) override;
+                void SwitchMultiSample(const uint16_t& newSampleCount, AZ::RPI::ViewportContextPtr viewportContext) override;
```
Personal preference: You dont need **&** for simple types like uint16_t

moudgils commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Reflect/System/RenderPipelineDescriptor.h:
```diff
@@ -18,6 +18,16 @@ namespace AZ
     namespace RPI
     {
+         // Rendering Settings
+        enum class AAMethod
```
Recommendation on name: AAMethod->AntiAliasingMode

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -74,6 +74,45 @@ void cvar_r_renderPipelinePath_Changed(const AZ::CVarFixedString& newPipelinePat
+    }
+    else
+    {
+        AZ_Warning("SetMultiSampleCount", false, "Failed to set multi sample count %d: invalid multi-sample count", newSampleCount);
```
Minor missed dash: to set multi **-** sample

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -74,6 +74,45 @@ void cvar_r_renderPipelinePath_Changed(const AZ::CVarFixedString& newPipelinePat
+    {
+        return;
+    }
+    if (newSampleCount > 0 && ((newSampleCount & (newSampleCount - 1))) == 0)
```
Add a comment about (newSampleCount & (newSampleCount - 1)) to ensure it is to check if the value is a power of 2. We should also probably clamp the value to be 16 for max value.

hosea1008 replied:
I clamp it up to 8, it seems that 16 is not acceptable by design [here](https://github.com/o3de/o3de/blob/66279081935e211510585a2fa725f1c2cc771cbb/Gems/Atom/Feature/Common/Code/Source/CoreLights/LightCullingTilePreparePass.cpp#L106)

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -83,6 +122,8 @@ AZ_CVAR(uint32_t, r_height, 1080, nullptr, AZ::ConsoleFunctorFlags::DontReplicat
 AZ_CVAR(uint32_t, r_resolutionMode, 0, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "0: render resolution same as window client area size, 1: render resolution use the values specified by r_width and r_height");
 AZ_CVAR(float, r_renderScale, 1.0f, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Scale to apply to the window resolution.");
+AZ_CVAR(AZ::CVarFixedString, r_antiAliasing, "", cvar_r_antiAliasing_Changed, AZ::ConsoleFunctorFlags::DontReplicate, "The anti-aliasing to be used for the current render pipeline. Available options: MSAA, TAA, SMAA");
+AZ_CVAR(uint16_t, r_multiSample, 0, cvar_r_multiSample_Changed, AZ::ConsoleFunctorFlags::DontReplicate, "The multi-sample count to be used for the current render pipeline."); // 0 stands for unchanged, load the default setting from the pipelien itself
```
Typo: pipeline

moudgils replied:
Recommendation: r_multiSample -> r_multiSampleCount

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -507,11 +566,16 @@ namespace AZ
 
+                    AZ::CVarFixedString antiAliasing = static_cast<AZ::CVarFixedString>(r_antiAliasing);
+                    if (antiAliasing != "")
+                        renderPipeline->SetActiveAAMethod(antiAliasing.c_str());
```
Needs {}

moudgils commented on Gems/Atom/Bootstrap/Code/Source/BootstrapSystemComponent.cpp:
```diff
@@ -564,6 +628,9 @@ namespace AZ
                 // it's setting the MSAA state from the last pipeline loaded.
+                const auto cvarMultiSample = static_cast<uint16_t>(r_multiSample);
+                if (cvarMultiSample > 0 && ((cvarMultiSample & (cvarMultiSample - 1))) == 0)
+                    multisampleState.m_samples = static_cast<uint16_t>(cvarMultiSample);
```
Needs {} as per code guidelines

moudgils commented on Gems/Atom/RPI/Code/Source/RPI.Public/RenderPipeline.cpp:
```diff
@@ -948,5 +950,96 @@ namespace AZ
+            }
+            return "MSAA";
+        }
+ 
```
Minor: Remote extra lines

moudgils commented on Gems/Atom/RPI/Code/Source/RPI.Public/RenderPipeline.cpp:
```diff
@@ -948,5 +950,96 @@ namespace AZ
+            }
+            return true;
+        }
+
```
Minor: Remove extra line

AMZN-alexpete commented on Gems/Atom/RPI/Code/Source/RPI.Public/RenderPipeline.cpp:
```diff
@@ -948,5 +950,93 @@ namespace AZ
+                return false;
+            }
+
+            static AZStd::unordered_map<AntiAliasingMode, Name> AAPassNamesLookup = {
+                {AntiAliasingMode::SMAA, Name("SMAA1xApplyLinearHDRColorPass")}, 
+                {AntiAliasingMode::TAA, Name("TaaPass")}
+            };
+
+            for (auto& aaPassMap : AAPassNamesLookup)
+            {
+                EnablePass(pipeline, aaPassMap.second, aaPassMap.first == aaMethod);
+            }
```
Don't we also need to enable the `ContrastAdaptiveSharpeningPass` when TAA is enabled? and disable it when not?

hosea1008 commented:
@AMZN-alexpete Hi, I think I have fixed the failed test case and rebased by the way, can you kindly help restarting the build and jenkins when you have time? Thanks!
